src/mlpack/methods/ne/genome.hpp

```diff
@@ -0,0 +1,329 @@
+  bool ContainEnabledLink(ssize_t innovId) const {
+    for(ssize_t i=0; i < NumLink(); ++i) {
+        if (aLinkGenes[i].InnovationId() == innovId &&
+            aLinkGenes[i].Enabled()) {
+            return true;
+        }
+    }
+    return false;
+  }
+
+  // Whether link exist.
+  bool ContainLink(ssize_t innovId) const {
+    for(ssize_t i=0; i < NumLink(); ++i) {
+        if (aLinkGenes[i].InnovationId() == innovId) {
+            return true;
+        }
+    }
+    return false;
+  }
+
+  // Set neurons' input and output to zero.
+  void Flush() {
+    for (ssize_t i=0; i<aNeuronGenes.size(); ++i) {
+      aNeuronGenes[i].Activation(0);
+      aNeuronGenes[i].Input(0);
+    }
+  }
+
+  // Sort link genes by toNeuron's depth.
+  void SortLinkGenes() {
+    struct DepthAndLink
+    {
+      double depth;
+      LinkGene link;
+ 
+      DepthAndLink(double d, LinkGene& l) : depth(d), link(l) {}
+
+      bool operator < (const DepthAndLink& dL) const
+      {
+        return (depth < dL.depth);
+      }
+    };
+
+    std::vector<double> toNeuronDepths;
+    for (ssize_t i=0; i<aLinkGenes.size(); ++i) {
+      NeuronGene toNeuron = GetNeuronById(aLinkGenes[i].ToNeuronId());
+      toNeuronDepths.push_back(toNeuron.Depth());
+    }
+
+    std::vector<DepthAndLink> depthAndLinks;
+    ssize_t linkGenesSize = aLinkGenes.size();
+    for (ssize_t i=0; i<linkGenesSize; ++i) {
+      depthAndLinks.push_back(DepthAndLink(toNeuronDepths[i], aLinkGenes[i]));
+    }
+
+    std::sort(depthAndLinks.begin(), depthAndLinks.end());
+
+    for (ssize_t i=0; i<linkGenesSize; ++i) {
+      aLinkGenes[i] = depthAndLinks[i].link;
+    }
+  }
+
+  // Activate genome. The last dimension of input is always 1 for bias. 0 means no bias.
+  void Activate(std::vector<double>& input) {
+    assert(input.size() == aNumInput);
+
+    SortLinkGenes();
+
+    // Set all neurons' input to be 0.
+    for (ssize_t i=0; i<NumNeuron(); ++i) {
+      aNeuronGenes[i].Input(0);
+    }
+
+    // Set input neurons.
+    for (ssize_t i=0; i<aNumInput; ++i) {
+      aNeuronGenes[i].Activation(input[i]);  // assume INPUT, BIAS, OUTPUT, HIDDEN sequence
+    }
+
+    // Activate hidden and output neurons.
+    for (ssize_t i = 0; i < NumLink(); ++i) {
+      if (aLinkGenes[i].Enabled()) {
+        ssize_t toNeuronIdx = GetNeuronIndex(aLinkGenes[i].ToNeuronId());
+        ssize_t fromNeuronIdx = GetNeuronIndex(aLinkGenes[i].FromNeuronId());
+        double input = aNeuronGenes[toNeuronIdx].Input() + 
+                       aNeuronGenes[fromNeuronIdx].Activation() * aLinkGenes[i].Weight();
+        aNeuronGenes[toNeuronIdx].Input(input);
+        
+        if (i == NumLink() - 1) {
+          aNeuronGenes[toNeuronIdx].CalcActivation();
+        } else if (GetNeuronIndex(aLinkGenes[i + 1].ToNeuronId()) != toNeuronIdx) {
+          aNeuronGenes[toNeuronIdx].CalcActivation();
+        }
+      }
+    }
+  }
+
+  // Get output vector.
+  std::vector<double> Output() {
```

Comment:
Let's use a reference here, also, it might be a good idea, to use an arma::vec here:

``
Output(std::vector<double>& output)
``